Add option for ignoring volumes defined in images

lorenz opened the pull request:
This adds a new config option for ignoring volumes defined in image metadata.

In Kubernetes volumes (even temporary ones) are generally specified in the pod configuration. Mapping the ones in the image metadata can lead to resource accounting/exhaustion issues (think writing an unlimited amount of inodes or data into the volume) and policy enforcement issues (a PSP cannot disable these).

Specifically for containers started with `ReadOnlyRootFilesystem` it can also lead to unexpected behavior, for example if there's a typo in the volume mount path the container will still start since the process can write to the data location, but the data ends up in the volume defined by the image which is ephemeral and subsequently gets lost when the pod gets descheduled. The user expected the process in the container to fail since there's no volume mounted at the data path and the root filesystem is readonly, but since the image volume is mounted there this works.

mikebrow commented:
/ok-to-test

mikebrow commented on pkg/server/container_create.go:
```diff
@@ -137,8 +137,11 @@ func (c *criService) CreateContainer(ctx context.Context, r *runtime.CreateConta
+	if !c.config.IgnoreImageDefinedVolumes {
+		// Create container image volumes mounts.
+		volumeMounts = c.volumeMounts(containerRootDir, config.GetMounts(), &image.ImageSpec.Config)
+	}
```
maybe some debug output when image spec config.Volumes is not nil and we are ignoring

mikebrow commented:
@lorenz sounds like this was a concrete issue you ran into. Can you provide link(s)?

lorenz commented:
I've added a debug log output if we're ignoring volumes and documented the option in config.md. I looked at the commit for the last-added option (TolerateMissingHugePagesCgroupController) and added mine to all the files it touched, but it wasn't added to `config.md`, that's why I missed it.

The main reason why I've been running this patch for quite some time is because of the issue where incorrectly configured volumes end up writing data into an ephemeral containerd-allocated volume and misbehaving containers writing lots of data into the containerd directory.

lorenz commented:
/test pull-cri-containerd-verify

mikebrow commented:
> I've added a debug log output if we're ignoring volumes and documented the option in config.md. I looked at the commit for the last-added option (TolerateMissingHugePagesCgroupController) and added mine to all the files it touched, but it wasn't added to `config.md`, that's why I missed it.
> 
> The main reason why I've been running this patch for quite some time is because of the issue where incorrectly configured volumes end up writing data into an ephemeral containerd-allocated volume and misbehaving containers writing lots of data into the containerd directory.

yeah I figured as much.. Which of course was my bad, either I missed it .. or just let it go in that pr review and flagged it as a docs issue to be worked on later, can't remember which was a week ago and seems like months :-)

mikebrow commented on pkg/server/container_create.go:
```diff
@@ -137,8 +137,13 @@ func (c *criService) CreateContainer(ctx context.Context, r *runtime.CreateConta
+	if !c.config.IgnoreImageDefinedVolumes {
+		// Create container image volumes mounts.
+		volumeMounts = c.volumeMounts(containerRootDir, config.GetMounts(), &image.ImageSpec.Config)
+	} else if len(image.ImageSpec.Config.Volumes) == 0 {
```
!=

lorenz replied:
Nice find, is now fixed

mikebrow commented:
> LGTM
> 
> @mikebrow Do we actually need the volume here? The user from eBay filed issue about this before [containerd/containerd#3511](https://github.com/containerd/containerd/issues/3511).
> 
> The volume handled by CRI plugin doesn't provide interface for user to handle this. The CRI-API spec defines volumes from kubelet, not image config. If the user uses devicemapper as snapshotter with disk quota, the size of rootfs will be ensured with quota. But volume from image config will be leak point.
> 
> cc @wenlxie

Hmm.  Reading the other issue (I re-assigned it to cri) it sounds like @Random-Liu already had some ideas for properly handling image specified volumes (report disk usage and /or ask kublet to manage these volumes). Maybe we can recruit him for a discussion at a future sig-node call.

Either way I think this flag provides a good ignore option for now at least until we have a more elaborate solution to map and/or report usage for the image volumes.

fuweid commented:
> Either way I think this flag provides a good ignore option for now at least until we have a more elaborate solution to map and/or report usage for the image volumes.

SGTM

lorenz commented:
Is this waiting on something to get merged?

mikebrow commented:
> Is this waiting on something to get merged?

Just giving maintainers a sufficient opportunity to weigh in.
